### PR TITLE
Rk4 solver

### PR DESCRIPTION
**Pull request by pmohtat**
# Description

Added the fixed-step Runge–Kutta method solver (RK4). The main motivation is that with a manual solver we can then build on with better solvers (which we can then export to C)

Fixes # (not an issue)

## Type of change

PR #s 995 was added to [CHANGELOG.md](https://github.com/pybamm-team/PyBaMM/blob/master/CHANGELOG.md) 

- [x] New feature (non-breaking change which adds functionality)
- [ ] Optimization (back-end change that speeds up the code)
- [ ] Bug fix (non-breaking change which fixes an issue)

# Key checklist:

- [x] No style issues: `$ flake8`
- [x] All tests pass: `$ python run-tests.py --unit`
- [x] The documentation builds: `$ cd docs` and then `$ make clean; make html`

You can run all three at once, using `$ python run-tests.py --quick`.

## Further checks:

- [x] Code is commented, particularly in hard-to-understand areas
- [x] Tests added that prove fix is effective or that feature works

### CODE REVIEW

**Comment by martinjrobins** [comment]
hi @pmohtat, thank you for the PR! Can you explain a bit more about the reasoning behind adding another implementation of RK4? the scipy solver in pybamm already includes this method. I'm all for implementing better solvers, but I'd be reluctant to add methods we already have

**Comment by pmohtat** [comment]
hi @martinjrobins, yes you are correct. But, the main motivation here was code generation (C++) for embedded controllers, simulink, etc. I wrote the integrator using CasAdi functions, so in theory we can use CasAdi's code generation to get the C code. Also, this way we are able to modify the solver freely and add needed functionality (e.g. adding algebraic equations to Rk4)

**Comment by martinjrobins** [comment]
OK, so the aim is to implement a faster version of rk4 than the scipy one, using casadi code generation? Is the code gen implemented in this pr? Be good to test if this is possible before we merge it in

**Comment by pmohtat** [comment]
we have tested the code generation on a separate project. I can add a code generation test in the test_rk4_solver.py.

**Comment by pmohtat** [comment]
hi. I have added a test in test_rk4_solver.py. I was able to compile and run the code both from command line and using Matlab's mex function generator.
Let me know if you need more information.

File: tests/unit/test_solvers/test_rk4_solver.py [hunk 1]
```diff
@@ -0,0 +1,247 @@
...
+        disc = pybamm.Discretisation()
+        model_disc = disc.process_model(model, inplace=False)
+        # Solve
+        solver = pybamm.RK4Solver(N=20)
+        t_eval = np.linspace(0, 1, 100)
+        solution = solver.solve(model_disc, t_eval)
+        opts = dict(main=True, mex=True)
+        solution.F.generate("gen.c", opts)
+        # To test the generated code run the following in the terminal
+        # gcc gen.c -o gen
```
Comment (martinjrobins) [inline tests/unit/test_solvers/test_rk4_solver.py:226]: if code generation is going to be a feature of this solver, how to do this should be documented as part of the class

File: pybamm/solvers/rk4_ode_solver.py [hunk 2]
```diff
@@ -0,0 +1,186 @@
...
+        solution.F = F
+        return solution
+
+    def get_integrator(self, model, t_eval, inputs):
+        # Only set up problem once
+        y0 = model.y0
+        if isinstance(y0, casadi.DM):
+            y0 = y0.full().flatten()
+        rhs = model.casadi_rhs
+        # algebraic = model.casadi_algebraic
```
Comment (martinjrobins) [inline pybamm/solvers/rk4_ode_solver.py:65]: what if this solver is used on a dae model? Should fail gracefully with some sort of error message

**Comment by pmohtat** [comment]
hi. thank you for the feedback, I'll hopefully start addressing them in coming days. I apologies for the many mistakes, I am pretty much a noob when it comes to pybamm and git...
I just want to reiterate that the motivation was code generation first and a new solver second. That's why I started with an easy to implement solver such as RK4.
If code generation is not a feature of interest l completely understand. Also, if you think there is better way to approach this without adding a custom solver let me know or point me in that direction. thanks again and happy to learn and contribute in any capacity!

**Comment by martinjrobins** [comment]
no worries at all, thank you for all your work contributing to pybamm! I think the code generation would be a good feature, but your solver still needs to conform to the interface established in `pybamm.BaseSolver`. Looking at the API docs (https://pybamm.readthedocs.io/en/latest/source/solvers/base_solver.html), you can see that the constructor takes `atol` and `rtol` as absolute and relative error tolerances to use for the solver. Therefore each solver must satisfy these error tolerances, whatever the specific time-points given to the solver.

File: pybamm/solvers/rk4_ode_solver.py [hunk 3]
```diff
@@ -0,0 +1,186 @@
...
+
+            solution.termination = "final"
+
+        return solution, F
+
+    def _run_integrator(self, integrator, model, y0, inputs, t_eval):
+        try:
+            sol = integrator
+            y_values = sol.y
+            return pybamm.Solution(sol.t, y_values)
```
Comment (tinosulzer) [inline pybamm/solvers/rk4_ode_solver.py:183]: I don't thing the `get_integrator` / `run_integrator` structure is appropriate in this case. Here, the integrator is not a function, but already the solution, and `get_integrator` and `run_integrator` are only called once each. So you could put the whole thing in the `_integrate` function

File: pybamm/solvers/rk4_ode_solver.py (hunk 2 above)
Comment (tinosulzer) [reply pybamm/solvers/rk4_ode_solver.py:65]: @pmohtat you can set `self.ode_solver=True` in the init function to do this check automatically, and remove this commented-out line

**Comment by martinjrobins** [comment]
Yes, I think this is not far off and will be a really good addition to the pybamm solvers. The main effort would be satisfying the tolerances, as this requires error estimation and interpolation to the points in `t_eval`. I've previously implemented a RK4(5) method with adaptive time-stepping and interpolation here (https://github.com/martinjrobins/swift-odeint). It's in a different language, but has citations you can read, and I'm happy to answer any questions you have about the algorithm. You can also have a look at the RK methods in scipy (https://github.com/scipy/scipy/blob/v1.4.1/scipy/integrate/_ivp/ivp.py#L157-L664), which have the advantage of being in python!

**Pull request closed without merging**